polus-graph-pyramid-builder-plugin/src/main.py

```diff
@@ -111,66 +103,373 @@ def bin_data(data,column_names):
+            datacol[ismin] = 1 # Smallest bin value is 1.
+            datacol[ismax] = bincount # Largest bin value is bincount
+            datacol[datacol == -np.inf] = negbin + 1
+            datacol[datacol == np.inf] = negbin + 1
+
+        commonratios = ratios.to_list()
+        alphas = alphavals.to_list()
+        yaxis = (negbins + 1).to_list()
+
+        return yaxis, alphas, commonratios, data
+
+    # Transforms the Data that has a Positive Range
+    def dfzero2pos(data, alphavals, datmax):
+
+        commonratios = (datmax/alphavals)**(1/(bincount - 1))
+
+        for col in data.columns: 
+
+            alpha = alphavals[col]
+            ratio = commonratios[col]
+
+            cols = data[col]
+
+            datacol = cols.to_numpy()
+            colname = cols.name
+
+            bin_data =  np.round(np.log(np.abs(datacol)/alpha)/np.log(ratio)) # Calculate the bin for all values
+            bin_data[bin_data==-0.0] = 0 # To replace -0.0 values with 0
+            ismax = (datacol == datmax[colname]) 
+
+            datacol[ismax] = bincount
+            datacol[bin_data<=0] = 1 # Negative bin value means that it was smaller than alpha
+            datacol[bin_data>0] = bin_data[bin_data>0] + 1
+            datacol[datacol == -np.inf] = 1
+            datacol[datacol == np.inf] = 1
+
+        alphas = alphavals.to_list()
+        commonratios = commonratios.to_list()
+
+        return alphas, commonratios, data
+
+    # Transform the Data that has a Negative Range
+    def dfneg2zero(data, datmin, alphavals):
+
+        commonratios = (datmin/alphavals)**(1/(bincount - 1))
+
+        for col in data.columns:
+            alpha = alphavals[col]
+            ratio = commonratios[col]
+
+            cols = data[col]
+            ogcol = cols.to_numpy()
+            datacol = cols.to_numpy()
+            colname = cols.name
+
+            bin_data =  np.round(np.log(datacol/alpha)/np.log(ratio)) # Calculate the bin for all values
+            bin_data[bin_data==-0.0] = 0 # To replace -0.0 values with 0
+
+            ismin = (datacol == datmin[colname]) 
+            datacol[ismin] = 1
+            datacol[bin_data<=0] = bincount -1 # Negative bin value means that it was bigger than alpha
+            datacol[bin_data>0] = bincount - bin_data[bin_data>0] + 1
+            datacol[datacol == -np.inf] = bincount 
+            datacol[datacol == np.inf] = bincount
+
+        alphas = alphavals.to_list()
+        commonratios = commonratios.to_list()
+
+        return alphas, commonratios, data
+
+    yaxis = [0]
+    column_bin_sizes = []
+    quartile25to75 = []
+    Datapoints = []
+    Ind2 = []
+    Ind2zero = []
+    Position = []
+    alphavals = []
+    
+    bin_stats = {'size': data.shape,
+                 'min': data.min(),
+                 'max': data.max(),
+                 'twenty5': data.quantile(0.25),
+                 'seventy5': data.quantile(0.75),
```

Comment:
Why not pass return these values and pass them to `binning_data'? That way the min and max are only calculated once and you don't have to recalculate in `binning_data` for log scaled data.

Reply:
So it gets recalculated for log scaled data because the columns in the log scaled graphs are different than the linear graphs. 
If the alpha value is zero, then those columns are not columns in the new data used to create the log scaled graphs.  
Otherwise, I would expect to get an IndexOutofBounds Error. 
Is there a different way I should go about this? 
Also why the shape is recalculated.

Reply:
I can also recalculate bin_stats outside this function as well.